llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp

```diff
@@ -0,0 +1,1410 @@
+                                             predecessors(&BB).end());
+      for (BasicBlock *Predecessor : Predecessors)
+        replaceBranchTargets(Predecessor, &BB, Successor);
+      ToRemove.push_back(&BB);
+    }
+
+    for (BasicBlock *BB : ToRemove)
+      BB->eraseFromParent();
+
+    return ToRemove.size() != 0;
+  }
+
+  bool addHeaderToRemainingDivergentDAG(Function &F) {
+    bool Modified = false;
+
+    auto MergeBlocks = getMergeBlocks(F);
+    auto ContinueBlocks = getContinueBlocks(F);
+    auto HeaderBlocks = getHeaderBlocks(F);
+
+    DomTreeBuilder::BBDomTree DT;
+    DomTreeBuilder::BBPostDomTree PDT;
+    PDT.recalculate(F);
+    DT.recalculate(F);
+
+    for (BasicBlock &BB : F) {
+      if (HeaderBlocks.count(&BB) != 0)
+        continue;
+      if (succ_size(&BB) < 2)
+        continue;
+
+      size_t CandidateEdges = 0;
+      for (BasicBlock *Successor : successors(&BB)) {
+        if (MergeBlocks.count(Successor) != 0 ||
+            ContinueBlocks.count(Successor) != 0)
+          continue;
+        if (HeaderBlocks.count(Successor) != 0)
+          continue;
+        CandidateEdges += 1;
+      }
+
+      if (CandidateEdges <= 1)
+        continue;
+
+      BasicBlock *Header = &BB;
+      BasicBlock *Merge = PDT.getNode(&BB)->getIDom()->getBlock();
+
+      bool HasBadBlock = false;
+      visit(*Header, [&](const BasicBlock *Node) {
+        if (DT.dominates(Header, Node))
+          return false;
+        if (PDT.dominates(Merge, Node))
+          return false;
+        if (Node == Header || Node == Merge)
+          return true;
+
+        HasBadBlock |= MergeBlocks.count(Node) != 0 ||
+                       ContinueBlocks.count(Node) != 0 ||
+                       HeaderBlocks.count(Node) != 0;
+        return !HasBadBlock;
+      });
+
+      if (HasBadBlock)
+        continue;
+
+      Modified = true;
+      Instruction *SplitInstruction = Merge->getTerminator();
+      if (isMergeInstruction(SplitInstruction->getPrevNode()))
+        SplitInstruction = SplitInstruction->getPrevNode();
+      BasicBlock *NewMerge =
+          Merge->splitBasicBlockBefore(SplitInstruction, "new.merge");
+
+      IRBuilder<> Builder(Header);
+      Builder.SetInsertPoint(Header->getTerminator());
+
+      auto MergeAddress = BlockAddress::get(NewMerge->getParent(), NewMerge);
+      SmallVector<Value *, 1> Args = {MergeAddress};
+      Builder.CreateIntrinsic(Intrinsic::spv_selection_merge, {}, {Args});
+    }
+
+    return Modified;
+  }
+
+public:
+  static char ID;
+
+  SPIRVStructurizer() : FunctionPass(ID) {
+    initializeSPIRVStructurizerPass(*PassRegistry::getPassRegistry());
+  };
+
+  virtual bool runOnFunction(Function &F) override {
+    bool Modified = false;
+
+    // In LLVM, Switches are allowed to have several cases branching to the same
+    // basic block. In SPIR-V, each target must be a distrinct block. This
+    // function makes sure each target is unique.
```

Comment:
I though you could repeat the same target bb multiple times in the same OpSwitch. See the code that glslang generates:

https://shader-playground.timjones.io/ba9ec8e0afb39a9fff42a1d0c0d6baa5

Reply:
If you need it for the algorithm, then that is something different.

Reply:
We can follow up in another PR.

Reply:
Yes, this constraint is only valid for if conditions. I'll update the comment to explain I also need this to avoid switches when not requires, as switch have a hard requirements for en OpSelectionMerge while if can avoid it in some cases.

Reply:
Ok turns out the structurizer tests using HLSL files were not running then executing `check-all` :upside_down_face: 
Running those removing this step makes some test break. Adding this again, but changing the comment to fix this bad argument.